README.md
Pants
=====
###### A Python3 implementation of the Ant Colony Optimization Meta-Heuristic

Overview
--------
**Pants** provides you with the ability to quickly determine how to visit a collection of interconnected nodes such that the work done is minimized. Nodes can be any arbitrary collection of data while the edges represent the amount of "work" required to travel between two nodes. Thus, **Pants** is a tool for solving traveling salesman problems.

The world is built from a list of edges. Edges are created from two nodes, and have a length that represents the amount of "work" in moving from the first node to the second node. Note that edge length need not represent the actual length of anything.  It could, for a silly, random example, be the number of dishes one must wash before moving to the next round of a dish-washing competition.

Solutions are found through an iterative process. In each iteration, several ants are allowed to find a solution that "visits" every node of the world. The amount of pheromone on each edge is updated according to its usefulness in finding shorter solutions. The ant that traveled the least distance is considered to be the local best solution. If the local solution has a shorter distance than the best from any previous iteration, it then becomes the global best solution. The elite ant(s) then deposit their pheromone along the path of the global best solution to strengthen it further, and the process repeats.

You can read more about [Ant Colony Optimization on Wikipedia](http://en.wikipedia.org/wiki/Ant_colony_optimization_algorithms).


Installation
------------
Currently there is no installation script (i.e., `setup.py`).  Simply copy the `pants/` directory to your `python` interpreter's `dist-packages/` directory:

```bash
$ sudo cp -r pants/ /usr/lib/python3/dist-packages
```


Useage
------
Using **Pants** is simple.  The example here uses Euclidean distance between 2D nodes with `(x, y)` coordinates, but there are no real requirements for node data of any sort.

 1) Import **Pants** (along with any other packages you'll need).

```python
import pants
import math
```

 2) Create `Node`s from your data points. Although the `Node` class is available for use, any *hashable* data type (such as `tuple` or `namedtuple`) will work.  `Node`s accept any keyword arguments and turns them into attributes. Here, `data_points` is a list of `dict`s.

```python
data_points = [
    {'x': 0, 'y': 0, 'name': 'origin'},
    {'x': 1, 'y': 1, 'name': 'node one'},
    {'x': 0, 'y': 5, 'name': 'node two'},
    {'x': 3, 'y': 4, 'name': 'node three'}
]
nodes = [pants.Node(**d) for d in data_points]
```

 3) Create `Edge`s and set their `length` property to represent the work required to traverse it.  Here the work required is the Euclidean distance between the two nodes (which have all been given `x` and `y` component properties to represent their position).

```python
edges = [Edge(a, b, length=math.sqrt(pow(a.x - b.x, 2) + pow(a.y - b.y, 2))]
```

 4) Create a `World` from the edges. Note that edges can also be added individually after the world has been instantiated by using the `add_edge` method.

```python
world = pants.World(edges[:-1])
world.add_edge(edges[-1])
```

 5) Create a `Solver` for the `World`.

```python
solver = pants.Solver(world)
```

 6) Solve the `World` with the `Solver`. Two methods are provided for finding solutions: `solve()` and `solutions()`. The former returns the best solution found, whereas the latter returns each solution found if it is the best thus far.

```python
solution = solver.solve()
# or
solutions = solver.solutions()
```

  7) Inspect the solution(s).

```python
print(solution.distance)
print(solution.path)
print(solution.moves)
# or
best = float("inf")
for solution in solutions:
  assert solution.distance < best
  best = solution.distance
```


Run the Demo
------------

Included is a 33 "city" demo that can be run from the command line.  Currently it accepts a single integer command line parameter to override the default iteration limit of 100.

```bash
$ cd Pants
$ ./bin/demo 100
Solver settings:
limit=100
rho=0.8, Q=1
alpha=1, beta=3
elite=0.5

Time Elapsed                Distance                 
--------------------------------------------------
           0:00:00.030429   0.7862956094256206       
           0:00:00.061907   0.7245780183747788       
           0:00:00.094099   0.6704966523088779       
           0:00:00.155262   0.649532279131667        
           0:00:00.425243   0.6478240330008148       
           0:00:00.486180   0.6460959831256239       
           0:00:00.998951   0.6386581061221168       
--------------------------------------------------
Best solution:
         0 = {"y": -84.221723, "x": 34.045483}
         1 = {"y": -84.225258, "x": 34.046006}
         4 = {"y": -84.224917, "x": 34.048679}
         8 = {"y": -84.226327, "x": 34.04951}
         9 = {"y": -84.218865, "x": 34.051529}
        14 = {"y": -84.217882, "x": 34.055487}
         5 = {"y": -84.216757, "x": 34.059412}
        12 = {"y": -84.217717, "x": 34.066471}
        20 = {"y": -84.225499, "x": 34.063814}
        30 = {"y": -84.22506, "x": 34.064489}
        19 = {"y": -84.242514, "x": 34.060164}
        29 = {"y": -84.243566, "x": 34.061518}
        10 = {"y": -84.240155, "x": 34.062461}
         6 = {"y": -84.237402, "x": 34.060461}
        28 = {"y": -84.255772, "x": 34.044915}
         2 = {"y": -84.262126, "x": 34.048194}
        27 = {"y": -84.267249, "x": 34.02115}
        22 = {"y": -84.363437, "x": 34.021342}
        25 = {"y": -84.36298, "x": 34.023101}
        23 = {"y": -84.36215, "x": 34.022585}
        24 = {"y": -84.361903, "x": 34.022718}
        21 = {"y": -84.33483, "x": 34.061468}
         7 = {"y": -84.334798, "x": 34.061281}
        16 = {"y": -84.283569, "x": 34.068647}
        15 = {"y": -84.283782, "x": 34.068455}
        13 = {"y": -84.265784, "x": 34.071628}
        11 = {"y": -84.21667, "x": 34.10584}
        17 = {"y": -84.177031, "x": 34.109645}
        31 = {"y": -84.163971, "x": 34.116852}
        18 = {"y": -84.163304, "x": 34.118162}
        26 = {"y": -84.16382, "x": 34.024302}
         3 = {"y": -84.208885, "x": 34.048312}
        32 = {"y": -84.20058, "x": 34.056326}
Solution length: 0.6386581061221168
Found at 0:00:00.998951 out of 0:00:02.994951 seconds.
$
```

Known Bugs
----------
 
None that I'm aware of currently.  Please let me know if you find otherwise!


Troubleshooting
---------------

Credits
-------
 
 * Robert Grant <rhgrant10@gmail.com>

License
-------

GPL


HISTORY.rst
=======
History
=======


0.7.0 (2020-04-18)
------------------

* Bump all dependencies to latest
* Fix minor display issue in the CLI
* Add py37 to the tox config


0.6.4 (2019-05-17)
------------------

* Fix the missing `acopy.utils` package problem


0.6.3 (2019-05-17)
------------------

* Freshen up the dev dependencies
* Add the Python 3.7 classifier
* Actually fix import issue


0.6.2 (2019-02-02)
------------------

* Fix import issue


0.6.1 (2018-10-07)
------------------

* Bump dependency on tsplib95 to 0.3.2


0.6.0 (2018-08-18)
------------------

* First release on PyPI as ``acopy``
* Complete rewrite
* Support for ``networkx``
* Support for ``tsplib95``
* Customizable solver
* Plotting capabilities
* Now uses Apache 2.0 License (formerly GPLv3)
* Supports only python 3.6+


0.5.2 (2014-09-09)
------------------

* Last release on the PyPI as ``ACO-Pants``


pants/pants.py
from math import sqrt
from datetime import timedelta
import itertools
import time
import random
import operator
import bisect

# TODO: Clean this up and put it somewhere else!
TEST_COORDS_5 = [
    (2, 2),
    (3, 1),
    (1, 2),
    (4, 3),
    (6, 1)
]

# TODO: Clean this up and put it somewhere else!
TEST_COORDS_33 = [
    (34.021150, -84.267249), (34.021342, -84.363437), (34.022585, -84.362150),
    (34.022718, -84.361903), (34.023101, -84.362980), (34.024302, -84.163820),
    (34.044915, -84.255772), (34.045483, -84.221723), (34.046006, -84.225258),
    (34.048194, -84.262126), (34.048312, -84.208885), (34.048679, -84.224917),
    (34.049510, -84.226327), (34.051529, -84.218865), (34.055487, -84.217882),
    (34.056326, -84.200580), (34.059412, -84.216757), (34.060164, -84.242514),
    (34.060461, -84.237402), (34.061281, -84.334798), (34.063814, -84.225499),
    (34.061468, -84.334830), (34.061518, -84.243566), (34.062461, -84.240155),
    (34.064489, -84.225060), (34.066471, -84.217717), (34.068455, -84.283782),
    (34.068647, -84.283569), (34.071628, -84.265784), (34.105840, -84.216670),
    (34.109645, -84.177031), (34.116852, -84.163971), (34.118162, -84.163304)
]

class Edge:
    """
    The connection between to coordinates.

    Each edge is composed of a distance and an amount of pheromone.

    """
    def __init__(self, a, b, dist=None, pheromone=0.1):
        """
        Create a new Edge between a and b.

        Parameters:
            a - the start of the edge
            b - the end of the edge
            dist - the distance between a and b (defaults to Euclidean)
            pheromone - the initial amount of pheromone (defaults to 0.1)

        """
        self.start = a
        self.end = b
        self.distance = Edge.distance(a, b) if dist is None else dist
        self.pheromone = 0.1 if pheromone is None else pheromone

    @staticmethod
    def distance(a, b):
        """
        Return the Euclidean distance between a and b.

        Parameters:
            a - the first point (x1, y1)
            b - the second point (x2, y2)
        Returns:
            sqrt((x2 - x1)^2 + (y2 - y1)^2)

        """
        x = b[0] - a[0]
        y = b[1] - a[1]
        return sqrt(x*x + y*y)
        
        
class Solver:
    """
    A world consisting of one or more coordinates in which ants find the
    shortest path that visits them all.

    """
    def __init__(self, world, **kwargs):
        self.world = world
        self.rho = kwargs.get('rho', 0.75)
        self.q = kwargs.get('Q', 1)
        self.t0 = kwargs.get('t0', .01)
        self.alpha = kwargs.get('alpha', 1.2)
        self.beta = kwargs.get('beta', 3)
        self.ant_count = kwargs.get('ant_count', 10)
        self.elite = kwargs.get('elite', .5)
        
    def solve(self, iter_count=20000):
        """
        Find the shortest path that visits every coordinate.
        """
        # Reset the amount of pheromone on every edge to the initial default.
        for edge in self.world.edges.values():
            edge.pheromone = self.t0

        # Yield local bests.
        # TODO: Add option to return global best.
        elite_ant = None
        for i in range(iter_count):
            # (Re-)Build the ant colony
            ants = self.round_robin_ants() if self.ant_count < 1 \
                    else self.random_ants()
            
            self.find_solutions(ants)
            self.update_scent(ants)
            best_ant = self.get_best_ant(ants)
            if elite_ant is None or best_ant < elite_ant:
                elite_ant = best_ant.clone()
            if self.elite:
                self.trace_elite(elite_ant)
            yield best_ant
    
    def round_robin_ants(self):
        n = len(self.world.coords)
        return [Ant(self.world, self.alpha, self.beta, start=self.world.coords[i % n])
            for i in range(self.ant_count)
        ]
        
    def random_ants(self):
        starts = self.world.coords[:]
        ants = list()
        while self.ant_count > 0 and len(starts) > 0:
            r = random.randrange(len(starts))
            ants.append(Ant(self.world, self.alpha, self.beta, start=starts.pop(r)))
        return ants

    def find_solutions(self, ants):
        """
        Let each ant find its way.
        """
        ants_done = 0
        while ants_done < len(ants):
            ants_done = 0
            for ant in ants:
                if ant.can_move():
                    m = ant.move()
                    self.world.edges[m].pheromone *= self.rho
                else:
                    ants_done += 1

    def update_scent(self, ants):
        """
        Update the amount of pheromone on each edge.
        """
        ants = sorted(ants)[:len(ants) // 2]
        for move, edge in self.world.edges.items():
            edge.pheromone = (1 - self.rho) * edge.pheromone + \
                    sum(self.q / a.distance for a in ants if move in a.moves)
            if edge.pheromone < self.t0:
                edge.pheromone = self.t0

    def get_best_ant(self, ants):
        """
        Return the ant with the shortest path.
        """
        return sorted(ants)[0]

    def trace_elite(self, ant):
        """
        Deposit pheromone along the path of a particular ant n times.
        """
        if self.elite:
            for m in ant.moves:
                self.world.edges[m].pheromone += self.elite * self.q / ant.distance
    

class World:
    def __init__(self, coords, edges=None):
        """
        Create a new world consisting of the given coordinates.

        The world is defined by a set of (x, y) coordinates, the assumption
        that each point can be reached from every other point, and a few
        other variables.

        Parameters:
            coords - list of (x, y) coordinates
            edges - dict of Edge instances with node-pair tuples as keys

        """
        self.coords = coords
        self.edges = edges or self.create_map()
        
    def create_map(self):
        """
        Create a map of the world from the coordinates.
        """
        edges = {}
        for a in self.coords:
            for b in self.coords:
                edges[a, b] = Edge(a, b)
                edges[b, a] = Edge(b, a, dist=edges[a, b].distance)
        return edges

    def distance(self, a, b):
        """
        Return the distance of the edge between a and b.
        """
        e = self.edges.get((a, b), None)
        return e.distance if e is not None else 0

    def scent(self, a, b):
        """
        Return the amount of pheromone on the edge between a and b.
        """
        e = self.edges.get((a, b), None)
        return e.pheromone if e is not None else 0


class Ant:
    """
    A single independent finder of solutions to the world.

    """
    uid = 0

    def __init__(self, world, alpha=1, beta=3, start=None):
        """
        Create a new Ant for the given world.

        Parameters:
            world - the World in which the ant should seek solutions
            alpha - how much this ant considers pheromone
            beta - how much this ant considers distance
            start - coordinate from which this ant should find solutions

        """
        Ant.uid = self.uid = Ant.uid + 1
        self.world = world
        self.alpha = alpha
        self.beta = beta
        self.trip_complete = False
        self.distance = 0
        self.reset(start)

    @property
    def moves(self):
        """
        A list of moves, where each move is a (start, end) coordinate tuple.
        """
        if len(self.path) == 0:
            return []
        path = self.path[:]
        starts = path[::2]
        path.append(path.pop(0))
        ends = path[::2]
        return zip(starts, ends)

    def clone(self):
        """
        Return a new ant with exactly the same property values as this ant.

        Note that unlike copy, this method preserves even the UID of an Ant.

        """
        ant = Ant(self.world, self.alpha, self.beta, self.start)
        ant.node = self.node
        ant.path = self.path[:]
        ant.distance = self.distance
        ant.trip_complete = self.trip_complete
        return ant

    def __lt__(self, other):
        return self.distance < other.distance

    def get_apriori(self, move):
        """
        Return a number suggesting how attractive a particular move seems.

        The default implementation uses inverse distance, but any apriori
        knowledge can be used instead.

        """
        if self.node is None:
            return 1
        return 1 / float(self.world.distance(self.node, move))

    def get_posteriori(self, move):
        """
        Return a number suggesting the amount of pheromone on the way to a move.
        """
        if self.node is None:
            return float(1)
        return self.world.scent(self.node, move)

    def log(self, msg):
        """
        Prints a message with the current timestamp and the ant's UID.
        """
        print("%s [Ant #%s] %s" % (time.time(), self.uid, msg))

    def reset(self, start=None):
        """
        Reset the ant so that it is ready to find another solution.

        Note that calling this method destroys the previous path, moves, and
        distance traveled by the ant.
        
        """
        self.start = start
        self.node = self.start
        self.distance = 0
        self.path = []
        self.trip_complete = False
        if start is not None:
            self.path.append(start)

    def can_move(self):
        """
        Return true if there is one or more coordinates not visited by the ant.
        """
        return not self.trip_complete

    def move(self):
        """
        Choose a valid move and make it.
        """
        moves = self.get_possible_moves()
        move = self.choose_move(moves)
        if move:
            move_made = (self.node, move)
            self.make_move(move)
            if len(self.path) == len(self.world.coords):
                # Close and complete the path.
                self.distance += self.world.distance(
                    self.path[-1], self.path[0]
                )
                self.trip_complete = True
            return move_made
        return None

    def get_possible_moves(self):
        """
        Return the set of all moves that can currently be made.
        """
        return set(self.world.coords) - set(self.path)

    def choose_move(self, moves):
        """
        Return the one move to make from a list of moves.

        The default implementation uses weighted probability based on edge
        distance and pheromone level.

        """
        if len(moves) == 0:
            return None     # No more moves
        
        # Find the individual weight of each move.
        moves = list(moves) # it may be given as a set, but we need order
        weights = [self.calculate_weight(m) for m in moves]
        
        # Normalize the weights without accedentally dividing by zero!
        total_weight = sum(weights)
        if total_weight == 0:
            total_weight = 1
        weights = [w / total_weight for w in weights]
        
        # Ensure the last element is 1 so that bisecting always returns a valid
        # index (instead of occasionally returning its length and producing an
        # error)
        cumdist = list(itertools.accumulate(weights)) + [1]
        
        # Choose a random place within the distrution of weights and return the
        # move at the index of would-be insertion.
        r = random.random()
        i = bisect.bisect(cumdist, r)
        return moves[i]
        
    def calculate_weight(self, move):
        """
        Return a number representing the weight of a single move.
        """
        p = self.get_posteriori(move)   # pheromone
        d = self.get_apriori(move)      # distance
        return pow(p, self.alpha) * pow(d, self.beta)

    def make_move(self, move):
        """
        Make the given move and update the distance traveled.
        """
        self.path.append(move)
        if len(self.path) == 1:
            self.start = move
        else:
            self.distance += self.world.distance(self.node, move)
        self.node = move

    
if __name__ == '__main__':
    world = World(TEST_COORDS_33)
    solver = Solver(world)
    print("""rho={w.rho}, Q={w.q}
alpha={w.alpha}, beta={w.beta}
elite={w.elite}""".format(w=solver))
    
    divider = "-" * (25 + 12 + 20)
    header = "\n{:21}{:12}{:20}".format("Time Elapsed", "Trial", "Distance")
    
    print(header)
    print(divider)
    fastest = None
    start_time = time.time()
    for i, ant in enumerate(solver.solve()):
        if fastest is None or ant.distance < fastest.distance:
            fastest = ant.clone()
            fastest_time = time.time() - start_time
            print("{:>20} {:<12}{:<20}".format(
                    timedelta(seconds=fastest_time), i, fastest.distance))
    total_time = time.time() - start_time
    
    print("\nTotal time for {} iterations: {}".format(
            i + 1,
            timedelta(seconds=total_time)))

    print(divider)
    print("Best solution:")
    for x, y in fastest.path:
        print("  {:>8} = ({:0.6f}, {:0.6f})".format(
                world.coords.index((x, y)),
                x,
                y))
    print("Time for best solution: {}".format(timedelta(seconds=fastest_time)))
